Add support for luks resource (#63)
community-terraform-providers/terraform-provider-ignition: website/docs/d/luks.html.md

* `name` - (Required) The name to use for the resulting luks device.

* `device` - (Required) The absolute path to the device. Devices are typically referenced by the /dev/disk/by-* symlinks.

* `label` - (Optional) The label of the luks device.

* `open_options` - (Optional) Any additional options to be passed to cryptsetup luksOpen. Supported options will be persistently written to the luks volume.

* `options` - (Optional) Any additional options to be passed to cryptsetup luksFormat.

* `uuid` - (Optional) The uuid of the luks device.

* `wipe_volume` - (Optional) Whether or not to wipe the device before luks creation.

* `key_file` - (Optional) Options related to the contents of the key file.

* `clevis` - (Optional) Describes the clevis configuration for the luks device.

The `key_file` block supports:

* `source` - (Required) The URL of the key file. Supported schemes are http, https, tftp, s3, arn, gs, and data. When using http, it is advisable to use the verification option to ensure the contents haven’t been modified.

* `compression` - (Optional) The type of compression used on the key file (null or gzip). Compression cannot be used with S3.

> * `httpHeaders` - (Optional) A list of HTTP headers to be added to the request. Available for http and https source schemes only.

Reviewer: ```suggestion
* `http_headers` - (Optional) A list of HTTP headers to be added to the request. Available for http and https source schemes only.
```

* `verification` - (Optional) Options related to the verification of the key file.

> 

Reviewer: Please add that `The http_headers blocks support:`...

The `verification` block supports:

> * `hash` - (Optional) The hash of the key file, in the form <type>-<value> where type is either sha512 or sha256. If compression is specified, the hash describes the decompressed key file.

Reviewer: There is no `verification.hash` defined in the schema. Please move the same description to `verification`.

Author: https://github.com/coreos/ignition/blob/main/config/v3_4/types/schema.go#L258

Reviewer: We do not have `hash` in our resource schema definition. See https://github.com/community-terraform-providers/terraform-provider-ignition/blob/v2/ignition/resource_ignition_config.go#L41-L45

Example doc: https://registry.terraform.io/providers/community-terraform-providers/ignition/latest/docs/data-sources/file#verification

Author: ah! good catch, let me fix that

The `clevis` block supports:

* `tang` - (Optional) describes a tang server. Every server must have a unique url.

* `tpm2` - (Optional) Whether or not to use a tpm2 device.

* `threshold` - (Optional) Sets the minimum number of pieces required to decrypt the device. Default is 1.

* `custom` - (Optional) Overrides the clevis configuration. The pin & config will be passed directly to clevis luks bind. If specified, all other clevis options must be omitted.

The `tang` block supports:

* `url` - (Optional) Url of the tang server.

* `thumbprint` - (Optional) Thumbprint of a trusted signing key.

* `advertisement` - (Optional) The advertisement JSON. If not specified, the advertisement is fetched from the tang server during provisioning.

The `custom` block supports:

* `pin` - (Optional) The clevis pin.

* `config` - (Optional) The clevis configuration JSON.

> * `needsNetwork` - (Optional) Whether or not the device requires networking.

Reviewer: ```suggestion
* `needs_network` - (Optional) Whether or not the device requires networking.
```

* `options` - (Optional) Any additional options to be passed to cryptsetup luksFormat.

* `uuid` - (Optional) The uuid of the luks device.

* `wipe_volume` - (Optional) Whether or not to wipe the device before luks creation.

* `key_file` - (Optional) Options related to the contents of the key file.

* `clevis` - (Optional) Describes the clevis configuration for the luks device.

The `key_file` block supports:

* `source` - (Required) The URL of the key file. Supported schemes are http, https, tftp, s3, arn, gs, and data. When using http, it is advisable to use the verification option to ensure the contents haven’t been modified.

* `compression` - (Optional) The type of compression used on the key file (null or gzip). Compression cannot be used with S3.

* `http_headers` - (Optional) A list of HTTP headers to be added to the request. Available for http and https source schemes only.

* `verification` - (Optional) The hash of the config, in the form _\<type\>-\<value\>_ where type is either sha512 or sha256. If compression is specified, the hash describes the decompressed file.

The `http_headers` block supports:

* `name` - (Required) The header name.

> * `value` - (Optional) The header contents.

Reviewer: ```suggestion
* `value` - (Required) The header contents.
```

Author: Note that the `value` of headers is optional in ignition, while declared as required in terraform's `httpHeaderReferenceResource`.

The `clevis` block supports:

* `tang` - (Optional) describes a tang server. Every server must have a unique url.

* `tpm2` - (Optional) Whether or not to use a tpm2 device.

* `threshold` - (Optional) Sets the minimum number of pieces required to decrypt the device. Default is 1.

* `custom` - (Optional) Overrides the clevis configuration. The pin & config will be passed directly to clevis luks bind. If specified, all other clevis options must be omitted.

The `tang` block supports:

> * `url` - (Optional) Url of the tang server.

Reviewer: ```suggestion
* `url` - (Required) Url of the tang server.
```
